# CS3033/CS6405 - Datamining - First Assignment

# Part 1 - Data Preparation








## 1.1. Load the dataset on Colab

In [1]:
# Importing pandas
import pandas as pd

# Reading in the file 
titanic = pd.read_csv("https://github.com/andvise/DataAnalyticsDatasets/blob/16ca8de1233c8643bfe85fcd1cd87c9ff2221312/titanic.csv?raw=True")

# Taking a look at the dataframe
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
882          883         0       2   
883          884         1       1   
884          885         0       3   
885          886         1       1   
886          887         0       3   

                                                  Name     Sex   Age  \
0                               Mr. Owen Harris Braund    male  22.0   
1    Mrs. John Bradley (Florence Briggs Thayer) Cum...  female  38.0   
2                                Miss. Laina Heikkinen  female  26.0   
3          Mrs. Jacques Heath (Lily May Peel) Futrelle  female  35.0   
4                              Mr. William Henry Allen    male  35.0   
..                                                 ...     ...   ...   
882                               Rev. Juozas Montvila    male  27.0   
883                        Miss. Margaret Edith Graham  female  19.0   
884                     Miss. Catherine Helen Johnston  female   7.0   
885                               Mr. Karl Howell Behr    male  26.0   
886                                 Mr. Patrick Dooley    male  32.0   

     Siblings/Spouses Aboard  Parents/Children Aboard     Fare  
0                          1                      0.0   7.2500  
1                          1                      0.0  71.2833  
2                          0                      0.0   7.9250  
3                          1                      NaN  53.1000  
4                          0                      0.0   8.0500  
..                       ...                      ...      ...  
882                        0                      0.0  13.0000  
883                        0                      0.0  30.0000  
884                        1                      2.0  23.4500  
885                        0                      0.0  30.0000  
886                        0                      0.0   7.7500  

[887 rows x 9 columns]


## 1.2. Display the attributes' name and their data type

In [2]:
# Most attributes are either of type 'int64' or 'float64' 
titanic.dtypes

PassengerId                  int64
Survived                     int64
Pclass                       int64
Name                        object
Sex                         object
Age                        float64
Siblings/Spouses Aboard      int64
Parents/Children Aboard    float64
Fare                       float64
dtype: object

## 1.3. Delete the columns PassengerId and Name

In [3]:
# Can remove these columns using the dataframe.drop() function from pandas
titanic = titanic.drop(columns = ['PassengerId', 'Name'])

titanic

Survived  Pclass     Sex   Age  Siblings/Spouses Aboard  \
0           0       3    male  22.0                        1   
1           1       1  female  38.0                        1   
2           1       3  female  26.0                        0   
3           1       1  female  35.0                        1   
4           0       3    male  35.0                        0   
..        ...     ...     ...   ...                      ...   
882         0       2    male  27.0                        0   
883         1       1  female  19.0                        0   
884         0       3  female   7.0                        1   
885         1       1    male  26.0                        0   
886         0       3    male  32.0                        0   

     Parents/Children Aboard     Fare  
0                        0.0   7.2500  
1                        0.0  71.2833  
2                        0.0   7.9250  
3                        NaN  53.1000  
4                        0.0   8.0500  
..                       ...      ...  
882                      0.0  13.0000  
883                      0.0  30.0000  
884                      2.0  23.4500  
885                      0.0  30.0000  
886                      0.0   7.7500  

[887 rows x 7 columns]

## 1.4. Replace all missing values with 0



In [4]:
# First check and see what attributes contain missing data
titanic.isnull().sum() 

Survived                   0
Pclass                     0
Sex                        0
Age                        0
Siblings/Spouses Aboard    0
Parents/Children Aboard    1
Fare                       0
dtype: int64

In [5]:
# Missing values can be set to 0 using the dataframe.fillna() function
titanic = titanic.fillna(0)

# Now there is no missing data
titanic.isnull().sum()

Survived                   0
Pclass                     0
Sex                        0
Age                        0
Siblings/Spouses Aboard    0
Parents/Children Aboard    0
Fare                       0
dtype: int64

## 1.5. Transform the Sex column into a numerical one

In [6]:
# Taking a look at the data in this attribute column:
titanic.Sex

0        male
1      female
2      female
3      female
4        male
        ...  
882      male
883    female
884    female
885      male
886      male
Name: Sex, Length: 887, dtype: object

In [7]:
# The Sex column data can be transformed to numerical data using pd.factorize()
# [0] specifies first part of .factorize() output (numerical data)
titanic.Sex = pd.factorize(titanic.Sex)[0]  

In [8]:
# Taking another look at the data in this attribute column
# male = 0 and female = 1
titanic.Sex

0      0
1      1
2      1
3      1
4      0
      ..
882    0
883    1
884    1
885    0
886    0
Name: Sex, Length: 887, dtype: int64


## 1.6. Use Survived as the target label and the rest of the data frame as features

Before dividing up the dataset, the data must first be shuffled randomly. This means that row indexes are no longer ordered from 0 to 887, but now have a random row order. This preprocessing step is performed in order to avoid potential biases that may have been introduced from the collection of data. 

In [9]:
# Shuffling the data - frac = 1 ensures the whole df is shuffled.
titanic = titanic.sample(frac = 1)

# the order of rows has now been randomized
titanic

Survived  Pclass  Sex   Age  Siblings/Spouses Aboard  \
91          0       1    0  46.0                        1   
656         0       1    0  58.0                        0   
184         0       1    0  39.0                        0   
44          0       3    0  30.0                        0   
2           1       3    1  26.0                        0   
..        ...     ...  ...   ...                      ...   
60          1       1    1  38.0                        0   
222         0       3    0  22.0                        0   
569         1       1    0  36.0                        0   
424         1       2    1  19.0                        0   
106         1       3    0  29.0                        0   

     Parents/Children Aboard      Fare  
91                       0.0   61.1750  
656                      2.0  113.2750  
184                      0.0   50.0000  
44                       0.0    8.0500  
2                        0.0    7.9250  
..                       ...       ...  
60                       0.0   80.0000  
222                      0.0    7.8958  
569                      0.0   26.3875  
424                      0.0   26.0000  
106                      0.0    7.7750  

[887 rows x 7 columns]

In [10]:
# Assigning Survived column as the target label
target_lab =  titanic.Survived

# 0 = not survived, 1 = survived
target_lab 

91     0
656    0
184    0
44     0
2      1
      ..
60     1
222    0
569    1
424    1
106    1
Name: Survived, Length: 887, dtype: int64

In [11]:
# Assigning the columns from PClass onwards as the features data
features = titanic.loc[:,"Pclass":]

features

Pclass  Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard  \
91        1    0  46.0                        1                      0.0   
656       1    0  58.0                        0                      2.0   
184       1    0  39.0                        0                      0.0   
44        3    0  30.0                        0                      0.0   
2         3    1  26.0                        0                      0.0   
..      ...  ...   ...                      ...                      ...   
60        1    1  38.0                        0                      0.0   
222       3    0  22.0                        0                      0.0   
569       1    0  36.0                        0                      0.0   
424       2    1  19.0                        0                      0.0   
106       3    0  29.0                        0                      0.0   

         Fare  
91    61.1750  
656  113.2750  
184   50.0000  
44     8.0500  
2      7.9250  
..        ...  
60    80.0000  
222    7.8958  
569   26.3875  
424   26.0000  
106    7.7750  

[887 rows x 6 columns]

## 1.7. Divide your dataset in 80% for training and 20% for test

Now the data can be split into training and test subsets.

80% of the feature and label data is assigned to training and 20% is assigned to testing.

In [12]:
# As 887 does not split evenly into 80/20 subsets, the cutoff value is rounded to the nearest whole value
# 709.6 -> 710
cutoff = round(len(titanic)*0.8) 

# ~80% of the data (710 rows)
train_features = features.iloc[:cutoff,]
train_labels = target_lab.iloc[:cutoff,]

# ~20% of data (177 rows)
test_features = features.iloc[cutoff:,]  
test_labels = target_lab.iloc[cutoff:,]

In [13]:
# Taking a look at the training features dataset (80% of original)
# Note that the indexes are still in the random order as before
train_features

Pclass  Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard  \
91        1    0  46.0                        1                      0.0   
656       1    0  58.0                        0                      2.0   
184       1    0  39.0                        0                      0.0   
44        3    0  30.0                        0                      0.0   
2         3    1  26.0                        0                      0.0   
..      ...  ...   ...                      ...                      ...   
438       3    0  20.0                        0                      0.0   
585       3    0  22.0                        0                      0.0   
580       1    0  36.0                        0                      0.0   
36        3    0  18.0                        0                      0.0   
730       2    0  23.0                        0                      0.0   

         Fare  
91    61.1750  
656  113.2750  
184   50.0000  
44     8.0500  
2      7.9250  
..        ...  
438    9.5000  
585    8.0500  
580   40.1250  
36     7.2292  
730   13.0000  

[710 rows x 6 columns]

In [14]:
# Taking a look at the training features dataset (20% of original)
test_features

Pclass  Sex    Age  Siblings/Spouses Aboard  Parents/Children Aboard  \
405       2    0   3.00                        1                      1.0   
792       1    1  49.00                        0                      0.0   
303       1    0   0.92                        1                      2.0   
15        2    1  55.00                        0                      0.0   
430       3    0  17.00                        0                      0.0   
..      ...  ...    ...                      ...                      ...   
60        1    1  38.00                        0                      0.0   
222       3    0  22.00                        0                      0.0   
569       1    0  36.00                        0                      0.0   
424       2    1  19.00                        0                      0.0   
106       3    0  29.00                        0                      0.0   

         Fare  
405   18.7500  
792   25.9292  
303  151.5500  
15    16.0000  
430    7.1250  
..        ...  
60    80.0000  
222    7.8958  
569   26.3875  
424   26.0000  
106    7.7750  

[177 rows x 6 columns]

## 1.8. Scale the columns using min-max scalers

In [15]:
# min_max_scaler function 
# Performs min-max normalization of df
def min_max_scaler(df):
    
    norm_df = df.copy()

    for col in range(len(norm_df.columns)):
        min_a = min(norm_df.iloc[:,col])  # min value of a specific column 
        max_a = max(norm_df.iloc[:,col])  # max value of a specific column

        for row in range(len(norm_df)):
            # min-max normalization of each data point in each column
            norm_df.iloc[row,col] = (norm_df.iloc[row,col] - min_a)/(max_a - min_a)
     
    return norm_df

In [ ]:
# Call this function on the training and testing feature data
sc_train_feat = min_max_scaler(train_features)
sc_test_feat = min_max_scaler(test_features)

In [17]:
# Min and max values of scaled training data is 0 and 1, respectively.
print(sc_train_feat.max())
print(sc_train_feat.min())

Pclass                     1.0
Sex                        1.0
Age                        1.0
Siblings/Spouses Aboard    1.0
Parents/Children Aboard    1.0
Fare                       1.0
dtype: float64
Pclass                     0.0
Sex                        0.0
Age                        0.0
Siblings/Spouses Aboard    0.0
Parents/Children Aboard    0.0
Fare                       0.0
dtype: float64


In [18]:
# Looking at the scaled training feature data as an example
# All values between 0 and 1
sc_train_feat

Pclass  Sex       Age  Siblings/Spouses Aboard  Parents/Children Aboard  \
91      0.0    0  0.572757                    0.125                 0.000000   
656     0.0    0  0.723549                    0.000                 0.333333   
184     0.0    0  0.484795                    0.000                 0.000000   
44      1.0    0  0.371701                    0.000                 0.000000   
2       1.0    1  0.321438                    0.000                 0.000000   
..      ...  ...       ...                      ...                      ...   
438     1.0    0  0.246042                    0.000                 0.000000   
585     1.0    0  0.271174                    0.000                 0.000000   
580     0.0    0  0.447097                    0.000                 0.000000   
36      1.0    0  0.220910                    0.000                 0.000000   
730     0.5    0  0.283740                    0.000                 0.000000   

         Fare  
91   0.119406  
656  0.221098  
184  0.097594  
44   0.015713  
2    0.015469  
..        ...  
438  0.018543  
585  0.015713  
580  0.078319  
36   0.014110  
730  0.025374  

[710 rows x 6 columns]

## 1.9 Print the shape of the train and test set

In [19]:
# Training set:
# Features dataframe containing 710 rows for 6 feature columns  
# Series of 710 target labels
print(sc_train_feat.shape, train_labels.shape)

# Testing set:
# Features dataframe containing 177 rows of data for 6 features
# Series of 177 target labels
print(sc_test_feat.shape, test_labels.shape)

(710, 6) (710,)
(177, 6) (177,)


# Part 2 - k-NN implementation

NOTE: My observed results may differ to other test runs due to randomization of the data at the start.

## 2.1 Defining Functions

All functions for K-NN implementation will be defined here before being called upon in later sections of Part 2.

In [20]:
# euclid_calc function
# Returns a dictionary containing the Euclidian distances of all training data 
# points from a specific row of testing query data.

def euclid_calc(query_index, train_df, test_df):

    euclids = {}
    # the row of testing data we are aiming to predict
    query_row = test_df.iloc[query_index,:]

    # calculate the Euclidian distance of all training data points
    for row in range(len(train_df)):
        # sum the squared distances
        sqrd_distance = 0
        for col in range(len(train_df.columns)):
            # Euclidian calculation
            calc = (train_df.iloc[row,col] - query_row[col])**2
            sqrd_distance += calc

        # get the square root of the sum of squared distances
        euclids[row] = sqrd_distance**0.5

    return euclids

In [21]:
# index_dict function
# Returns a dictionary containing an enumerated version of the original randomized indexes
# e.g. if the first two rows of a df were indexes 122 and 450 after randomization, 
# the dictionary will be {0:122, 1:450....}
# Required for keeping track of what values the randnomized index numbers actually correspond to.

def index_dict(df):
    
    row_inds = {}
    for i in range(len(df.index)):
        row_inds[i] = df.index[i]
    
    return row_inds

In [22]:
# knn_fit function
# Computes the predicted results from a dictionary of Euclidian distances.
# Extracts k points with the smallest distance to the query point.
# Selects the most common class among the k points as the prediction value.
# Returns a dictionary containing the actual and predicted target label of each row.
# Default k set to 5, but can be changed.

def knn_fit(train_feat, test_feat, train_labels, test_labels, euclid_dists, k=5):
    
    # To keep track of train and test features indexing
    train_ind = index_dict(train_feat)
    test_ind = index_dict(test_feat)

    results = {}    
    for d in range(len(euclid_dists)):

        # In the results dictionary, assign the row number of the test data being predicted 
        # to another dictionary containg the actual test label for this specific row.
        results['Test row ' + str(test_ind[d])] = {'Actual':test_labels[test_ind[d]]}

        # k nearest neighbours found by sorting the dictionary 
        k_nearest = sorted(euclid_dists[d].values())[:k]
        
        # Store classes of the k nearest neighbours
        classes = []
        for key, value in euclid_dists[d].items():
            
            if value in k_nearest:
                # correct index of training row corresponding to k nearest value
                index = train_ind[key]
                # append 1 or 0 label to classes list
                classes.append(train_labels[index])
   
        # count classes in the list and assign prediction as most common label
        if classes.count(1) > classes.count(0):
            results['Test row ' + str(test_ind[d])]['Prediction'] = 1 
        else:
            results['Test row ' + str(test_ind[d])]['Prediction'] = 0
        
    return results

In [23]:
# evaluate_knn function
# Computes the accuracy of a particular K-NN model.
# Takes the K-NN results dictionary as input.

def evaluate_knn(result_dict):
    
    score = 0
    for d in result_dict.values():
        # if prediction correct
        if d['Actual'] == d['Prediction']:
            score +=1
    
    # correct predictions over total predictions
    accuracy = score/len(result_dict)
    
    return accuracy


# confusion_matrix function
# Generates confusion matrix from the K-NN results dictionary as input.  
# Confusion matrix/Table visualization code inspired from 
# https://pandas.pydata.org/pandas-docs/stable/user_guide/style.html 
# https://www.statology.org/confusion-matrix-python/

def confusion_matrix(result_dict):
    
    actual = []
    predicted = []
    
    for d in result_dict.values():
        actual.append(d['Actual'])        # Actual test label
        predicted.append(d['Prediction']) # Predicted test label
      
    # Make each list into a series
    actual = pd.Series(actual, name='Actual')
    predicted = pd.Series(predicted, name='Predicted')

    # Use crosstab function to get contingency table/confusion matrix
    conf_mat = pd.crosstab(actual, predicted)
    
    # Formatting the confusion matrix
    conf_mat = conf_mat.style.background_gradient(cmap ='Blues')\
                .set_properties(**{'font-size': '35px'})
    
    return conf_mat

## 2.2 Euclidian Distances

For each of the rows in the testing dataset, the Euclidian distances are calculated between all training dataset rows and the testing data row being queried. Each dictionary of Euclidian distances for each query row is then stored in a new dictionary, reuslting in a dictioanry of dictionaries. 

This ***euclid_calc*** function was made separate to the ***knn_fit*** function as it takes some time (~30s) to compute. Once calculated, this euclidian distances dictionary can be accessed and used as input for other functions, alleviating the need for repeating the computation if a new k value was to be tested.

In [24]:
euclid_dists = {}

# Iterating through each row of the scaled testing data
for i in range(len(sc_test_feat)):
    euclid_dists[i] = euclid_calc(i, sc_train_feat, sc_test_feat)

C:\Users\conor\AppData\Local\Temp\ipykernel_20752\651777586.py:17: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  calc = (train_df.iloc[row,col] - query_row[col])**2



The first key of this new dictionary is the reference index* (0-177) of the testing row in question, with the value of each of these keys being the dictionary of Euclidian distances calculated for all 710 training rows. So the euclid_dists dictionary contains 177 keys, each of which contain a dictionary of 710 keys (training row index) with 710 values (Euclidian distances). An example subset of this dictionary can be seen below. This is the first dictionary of Euclidian distances i.e. the Euclidian distances between all training row data and the first row of testing data.


In [25]:
euclid_dists[0]

{0: 0.7676703903442695,
 1: 0.8920409925196864,
 2: 0.7172200489792645,
 3: 0.6492187409007891,
 4: 1.1789267578257463,
 5: 1.1512214753868275,
 6: 0.5979259132223799,
 7: 1.1514087059340439,
 8: 1.2116202320538965,
 9: 0.7478698953985591,
 10: 0.6186811250409844,
 11: 0.718454151317223,
 12: 0.7314202660496374,
 13: 0.5932183336469747,
 14: 0.5978139766026666,
 15: 0.4667851335696587,
 16: 1.347396778235987,
 17: 0.8281568338640459,
 18: 0.5607476353938214,
 19: 1.1583934469018868,
 20: 0.8824513059511043,
 21: 0.6697341648931932,
 22: 1.1692416719861858,
 23: 0.5806235341918573,
 24: 0.9155936677593701,
 25: 1.2009730497217288,
 26: 1.0198825195403842,
 27: 1.0745834889897843,
 28: 0.7938569912373719,
 29: 0.7309642390411866,
 30: 1.1604721949240526,
 31: 0.3450651010454138,
 32: 0.6174950944376515,
 33: 1.224933211243321,
 34: 1.08220827539334,
 35: 1.0501887092011504,
 36: 0.8929298079873467,
 37: 1.1256750613215707,
 38: 0.7987640505425359,
 39: 0.03818283863166163,
 40: 0.6697266

#### Indexing
To take a closer look at an example of the reference indexing, we can call the ***index_dict*** function on the scaled testing feature data. Each key is ordered from 0-177, but the actual true row index of the dataset at each key index is different. The actual true row index of the randomized data corresponds to the values of these 177 keys. This is why I refer to it as a 'reference'. This can be seen below:

In [26]:
# Take note of the indexes of the testing data
sc_test_feat 

Pclass  Sex       Age  Siblings/Spouses Aboard  Parents/Children Aboard  \
405     0.5    0  0.031961                    0.125                      0.2   
792     0.0    1  0.738783                    0.000                      0.0   
303     0.0    0  0.000000                    0.125                      0.4   
15      0.5    1  0.830977                    0.000                      0.0   
430     1.0    0  0.247081                    0.000                      0.0   
..      ...  ...       ...                      ...                      ...   
60      0.0    1  0.569760                    0.000                      0.0   
222     1.0    0  0.323909                    0.000                      0.0   
569     0.0    0  0.539029                    0.000                      0.0   
424     0.5    1  0.277812                    0.000                      0.0   
106     1.0    0  0.431469                    0.000                      0.0   

         Fare  
405  0.036598  
792  0.050610  
303  0.295806  
15   0.031230  
430  0.013907  
..        ...  
60   0.156150  
222  0.015412  
569  0.051505  
424  0.050749  
106  0.015176  

[177 rows x 6 columns]

In [27]:
# Keys correspond to the ordered number of rows in the dataset.
# Their associated values correspond to the actual true indexes seen above.
# This structure allows the true indexing values to be accessed easily within the for loops of functions

test_inds = index_dict(sc_test_feat)
test_inds

{0: 405,
 1: 792,
 2: 303,
 3: 15,
 4: 430,
 5: 673,
 6: 803,
 7: 363,
 8: 445,
 9: 710,
 10: 336,
 11: 341,
 12: 865,
 13: 787,
 14: 776,
 15: 163,
 16: 554,
 17: 800,
 18: 328,
 19: 156,
 20: 592,
 21: 251,
 22: 231,
 23: 433,
 24: 3,
 25: 89,
 26: 143,
 27: 364,
 28: 255,
 29: 48,
 30: 745,
 31: 346,
 32: 530,
 33: 764,
 34: 356,
 35: 699,
 36: 58,
 37: 693,
 38: 688,
 39: 493,
 40: 815,
 41: 283,
 42: 659,
 43: 526,
 44: 635,
 45: 37,
 46: 826,
 47: 11,
 48: 199,
 49: 647,
 50: 628,
 51: 331,
 52: 46,
 53: 71,
 54: 183,
 55: 263,
 56: 817,
 57: 168,
 58: 278,
 59: 419,
 60: 721,
 61: 506,
 62: 683,
 63: 40,
 64: 80,
 65: 33,
 66: 450,
 67: 132,
 68: 695,
 69: 120,
 70: 261,
 71: 31,
 72: 718,
 73: 681,
 74: 398,
 75: 833,
 76: 886,
 77: 114,
 78: 472,
 79: 846,
 80: 437,
 81: 503,
 82: 420,
 83: 864,
 84: 788,
 85: 373,
 86: 246,
 87: 259,
 88: 485,
 89: 842,
 90: 332,
 91: 860,
 92: 784,
 93: 142,
 94: 870,
 95: 527,
 96: 126,
 97: 759,
 98: 816,
 99: 733,
 100: 544,
 101: 417,
 1

## 2.3 K-NN Predictions 

Extracting the results of the K-NN model using the ***knn_fit*** function previously defined:

In [28]:
knn_results = knn_fit(sc_train_feat, sc_test_feat, train_labels, test_labels, euclid_dists, k=5)

The results are a dictionary object containing the actual target label and the predicted target label for each row of the testing data. Note that in Google Colab, the dictionary keys (strings here) are **ordered** as default. All row indexes do correspond to the true row index from the randomized test feature data, but they just appear ordered here.

(In Jupyter notebook the keys will appear as they are added and will not be ordered, meaning the order of keys would correspond direclty to the order of rows in the randomized data, i.e. the first row in the dictionary would be 'Test Row ...[first true index of feature data ]')

In [29]:
knn_results  # dictionaries odered by default 

{'Test row 405': {'Actual': 1, 'Prediction': 1},
 'Test row 792': {'Actual': 1, 'Prediction': 1},
 'Test row 303': {'Actual': 1, 'Prediction': 1},
 'Test row 15': {'Actual': 1, 'Prediction': 1},
 'Test row 430': {'Actual': 0, 'Prediction': 0},
 'Test row 673': {'Actual': 0, 'Prediction': 1},
 'Test row 803': {'Actual': 0, 'Prediction': 0},
 'Test row 363': {'Actual': 0, 'Prediction': 0},
 'Test row 445': {'Actual': 1, 'Prediction': 1},
 'Test row 710': {'Actual': 0, 'Prediction': 0},
 'Test row 336': {'Actual': 1, 'Prediction': 0},
 'Test row 341': {'Actual': 0, 'Prediction': 0},
 'Test row 865': {'Actual': 1, 'Prediction': 1},
 'Test row 787': {'Actual': 0, 'Prediction': 0},
 'Test row 776': {'Actual': 1, 'Prediction': 1},
 'Test row 163': {'Actual': 0, 'Prediction': 0},
 'Test row 554': {'Actual': 0, 'Prediction': 0},
 'Test row 800': {'Actual': 1, 'Prediction': 0},
 'Test row 328': {'Actual': 1, 'Prediction': 1},
 'Test row 156': {'Actual': 0, 'Prediction': 0},
 'Test row 592': {'Ac

## 2.4 Evaluating the K-NN model

Before computing the testing accuracy, we can calculate the baseline accuracy that can be achieved with this data. If we tally both classes (survived or not) and divide the max tally by the total number of classes, we can identify the minimum accuracy that can be achieved. In my randomized data, the baseline accuracy was 0.58192.

In [30]:
most_common = max([test_labels[test_labels==1].count(),test_labels[test_labels==0].count()])

baseline_acc = most_common/len(test_features)

baseline_acc

0.615819209039548

Evaluating the actual testing accuracy of this K-NN model using the ***evaluate_knn*** function previosuly defined.In my testing of the model, the model achieved an accuracy of 0.8248, which is a reasonably good level of accuracy.

In [31]:
evaluate_knn(knn_results)

0.8418079096045198

Plotting the confusion matrix using the ***confusion_matrix*** function previously defined:

In [32]:
confusion_matrix(knn_results)

The 177 cases from the testing feature data are represented in this confusion matrix. The rows correspond to the actual target label for these cases, while columns correspond to the predicted target label. A value of 0 means the person did not survive, while a value of 1 means they did survive. 

***Moving across the matrix:***

The first cell of the matrix [0,0] represents the number of cases the model predicted as 'not survived' when the actual target label was 'not survived'. This corresponds to a correct prediction. This number is the highest of all cells in the matrix, highlighting the model's superior ability to accurately predict those who did not survive.

The next cell of the matrix [0,1] represents the number of cases the model predicted as 'survived' when the actual target label was 'not survived'. This corresponds to an incorrect prediction.

The next cell of the matrix [1,0] represents the number of cases the model predicted as 'not survived' when the actual target label was 'survived'. This corresponds to an incorrect prediction. This number is the largest of both incorrect predictions in the matrix, meaning when the model was incorrect in its prediction, it was more likely predicting those who 'survived' as 'not survived'.

The last cell of the matrix [1,1] represents the number of cases the model predicted as 'survived' when the actual target label was 'survived'. This corresponds to a correct prediction. This number is significantly less than the other correctly predicted cases in the matrix [0,0], again highlighting the fact that the model was more capable of predicting those who did not survive.

**Note: values may differ across various test runs of the code.**


# Part 3 - Hyperparameters search 


Testing possible k values:

In [33]:
hyperpar_res = []
possible_k = [1, 3, 5, 7, 9, 11]

for kval in possible_k:
    
    k_result = knn_fit(sc_train_feat, sc_test_feat, train_labels, test_labels, euclid_dists, k=kval)
    
    hyperpar_res.append(k_result)

for d in range(len(hyperpar_res)):
    print('%f testing accuracy with k = %i' % (evaluate_knn(hyperpar_res[d]), possible_k[d]))
    

0.824859 testing accuracy with k = 1
0.830508 testing accuracy with k = 3
0.841808 testing accuracy with k = 5
0.807910 testing accuracy with k = 7
0.824859 testing accuracy with k = 9
0.813559 testing accuracy with k = 11


Confusion matrix of results generated from K-NN with k = 5:

In [34]:
confusion_matrix(hyperpar_res[3])

Confusion matrix of results generated from K-NN with k = 11:

In [35]:
confusion_matrix(hyperpar_res[5])

In my particular testing of the model, the best accuracy value was achieved with a k value of 5 or 11 (tied value). From observing both confusion matrices, the performance of each is almost identical. In this case I would choose the lower k value of 5, as it achieves the same accuracy as k = 11. A k value of 5 appears to be large enough so as to avoid the influence of potential outliers that may present nearer to the target point. In addition, a k value of 5 is small enough so that it is not incorporating too many 'nearest' points, especially those that may actually be closer to other targets. 

In this test run, each of the k values (excluding 1) appear to perform quite similarly, meaning the optimal value of k in this case is only marginally improving on the other k values. Despite generating an accuracy of 0.774, it is clear that a k value of 1 is the worst choice of k, as it is just too small and is only considering one nearest neighbour, which is not enough to base an accurate prediction on. 

# Part 4 - Weighted k-NN 




Function for performing weighted k-NN. The distance weighted metric in this case is the inverse distance squared.

As we are dealing with a binary classification problem (0 or 1 label), we can just sum the weighted distances for each target label individually. The target label with the largest sum of weights will be the overall precition for a target row of testing data.

Also, as it is a weighted k-NN being performed, I have included an option for using 'all' points in the dataset instead of just k nearest neighbours. Points that are nearer to the target will be weighted the highest, while those further away will have very small weights, and little to no influence on the predicted outcome. Therefore, technically all points can be involved.

***Accounting for 0 distance:***

To account for two points potentially having 0 distance and to avoid a division by zero error, I have added a miniscule value (1e-7) onto the distance squared value within the weight computation. This will prevent any value of zero from occuring and thus there should be no division by zero.

In [36]:
# weighted_knn function
# Similar to knn_fit function from before
# Computes the predicted results from a dictionary of Euclidian distances.
# Calculates the inverse Euclidian distance squared.
# Sums these weighted distances for both target labels.
# Label corresponding to the maximum sum of weights = the predicted target label.
# Returns a dictionary containing the actual and predicted target label of each row.
# Default k set to 5, but can be changed and also set to 'all'.

def weighted_knn(train_feat, test_feat, train_labels, test_labels, euclid_dists, k=5):
    
    train_ind = index_dict(train_feat)
    test_ind = index_dict(test_feat)
    
    w_knn = {}

    for i in range(len(euclid_dists)):

        # for summing the weighted distances of each target label
        surv_sum = 0
        deceased_sum = 0
    
        w_knn['Test row ' + str(test_ind[i])] = {'Actual':test_labels[test_ind[i]]}

        # if all points are selected instead of just k nearest
        if k == 'all':
            k_nearest = euclid_dists[i].values()
        else:
            k_nearest = sorted(euclid_dists[i].values())[:k]
        
        for key, value in euclid_dists[i].items():
    
            if value in k_nearest:

                index = train_ind[key]

                # calculate the inverse weighted square + 1e-7
                w = 1/(value**2 + 0.0000001)         

                # if label is 1, add to survived sum of weights
                # otherwise add to deceased
                if train_labels[index] == 1:
                    surv_sum += w
                else:
                    deceased_sum += w

        # same as knn_fit
        if surv_sum > deceased_sum:
            w_knn['Test row ' + str(test_ind[i])]['Prediction'] = 1
        else:
            w_knn['Test row ' + str(test_ind[i])]['Prediction'] = 0
            
    return w_knn

In [37]:
w_knn = weighted_knn(sc_train_feat, sc_test_feat, train_labels, test_labels, euclid_dists, k=5)

# similar output as before
w_knn

{'Test row 405': {'Actual': 1, 'Prediction': 1},
 'Test row 792': {'Actual': 1, 'Prediction': 1},
 'Test row 303': {'Actual': 1, 'Prediction': 1},
 'Test row 15': {'Actual': 1, 'Prediction': 0},
 'Test row 430': {'Actual': 0, 'Prediction': 0},
 'Test row 673': {'Actual': 0, 'Prediction': 1},
 'Test row 803': {'Actual': 0, 'Prediction': 0},
 'Test row 363': {'Actual': 0, 'Prediction': 0},
 'Test row 445': {'Actual': 1, 'Prediction': 1},
 'Test row 710': {'Actual': 0, 'Prediction': 0},
 'Test row 336': {'Actual': 1, 'Prediction': 0},
 'Test row 341': {'Actual': 0, 'Prediction': 0},
 'Test row 865': {'Actual': 1, 'Prediction': 1},
 'Test row 787': {'Actual': 0, 'Prediction': 0},
 'Test row 776': {'Actual': 1, 'Prediction': 1},
 'Test row 163': {'Actual': 0, 'Prediction': 0},
 'Test row 554': {'Actual': 0, 'Prediction': 0},
 'Test row 800': {'Actual': 1, 'Prediction': 0},
 'Test row 328': {'Actual': 1, 'Prediction': 1},
 'Test row 156': {'Actual': 0, 'Prediction': 0},
 'Test row 592': {'Ac

Evaluating the accuracy of weighted k-NN with different values for k, and comapring these values to the accuracy achieved with the original k-NN model.

In [38]:
weight_res = []
original_res = []
possible_k = [1, 3, 5, 7, 9, 11, 13, 15] # try some new k values

for k in possible_k:
    
    # compute original and weighted results
    wknn_result = weighted_knn(sc_train_feat, sc_test_feat, train_labels, test_labels, euclid_dists, k=k)
    knn_result = knn_fit(sc_train_feat, sc_test_feat, train_labels, test_labels, euclid_dists, k=k)
    
    weight_res.append(wknn_result)
    original_res.append(knn_result)

# evaluate both original and testing accuracy 
for d in range(len(weight_res)):
    print('%f Weighted testing accuracy with k = %s' % (evaluate_knn(weight_res[d]), str(possible_k[d])))
    print('%f Original testing accuracy with k = %i' % (evaluate_knn(original_res[d]), possible_k[d]))
    print()

# testing all points as k in the weighted model
wknn_all = weighted_knn(sc_train_feat, sc_test_feat, train_labels, test_labels, euclid_dists, k='all')
print('%f Weighted testing accuracy with k = all' % (evaluate_knn(wknn_all)))

0.824859 Weighted testing accuracy with k = 1
0.824859 Original testing accuracy with k = 1

0.813559 Weighted testing accuracy with k = 3
0.830508 Original testing accuracy with k = 3

0.824859 Weighted testing accuracy with k = 5
0.841808 Original testing accuracy with k = 5

0.824859 Weighted testing accuracy with k = 7
0.807910 Original testing accuracy with k = 7

0.836158 Weighted testing accuracy with k = 9
0.824859 Original testing accuracy with k = 9

0.830508 Weighted testing accuracy with k = 11
0.813559 Original testing accuracy with k = 11

0.836158 Weighted testing accuracy with k = 13
0.813559 Original testing accuracy with k = 13

0.836158 Weighted testing accuracy with k = 15
0.790960 Original testing accuracy with k = 15

0.807910 Weighted testing accuracy with k = all


In my test run of the model:

The weighted k-KNN model does not outperform the original model when both have k = 5. However, when k=11 the weighted k-NN model outperforms the original model at k = 7, 9, 11, 13 and 15, suggesting that the weighted model performs slightly better than the original at higher values of k. 

The maximum accuracy was achieved with the weighted model, with a value of 0.830508 at k = 11, 13 and 15. The testing accuracy evidently plateaud across these values, but could maybe be slighlty improved upon at higher values.

Interestingly, when incorporating all values in the dataset as k, the accuracy decreases to 0.813559. 

***Again note these results may differ across different test runs due to randomization of the data at the start. The model does not take long to run so one can run it again if they please and check other possible results.***